service/storage_service.cc

```diff
@@ -1906,48 +1909,174 @@ future<> storage_service::do_stop_ms() {
-        int nr_seconds = 10;
-        while (!(*heartbeat_updater_done) && nr_seconds--) {
-            co_await sleep(std::chrono::seconds(1));
+        co_await coroutine::return_exception_ptr(std::move(ex));
+    }
+
+    future<> send_to_all(node_ops_cmd cmd) {
+        req.cmd = cmd;
+        sstring op_desc = format("{}", cmd);
+        bool is_prepare = op_desc.find("prepare") != sstring::npos;
+        slogger.info("{}[{}]: Started {}, sync_nodes={}", desc, uuid(), op_desc, sync_nodes);
+        co_await coroutine::parallel_for_each(sync_nodes, [&] (const gms::inet_address& node) -> future<> {
+            if (nodes_unknown_verb.contains(node) || nodes_down.contains(node) ||
+                    (nodes_failed.contains(node) && op_desc != "abort")) {
+                // Note that we still send to failed nodes that 
+                co_return;
+            }
+            try {
+                co_await ss._messaging.local().send_node_ops_cmd(netw::msg_addr(node), req);
+                slogger.debug("{}[{}]: Got {} response from node={}", desc, uuid(), op_desc, node);
+            } catch (const seastar::rpc::unknown_verb_error&) {
+                if (is_prepare) {
+                    slogger.warn("{}[{}]: Node {} does not support the {} verb", desc, uuid(), node, op_desc);
+                } else {
+                    slogger.warn("{}[{}]: Node {} did not find ops_uuid={} or does not support the {} verb", desc, uuid(), node, uuid(), op_desc);
+                }
+                nodes_unknown_verb.emplace(node);
+            } catch (const seastar::rpc::closed_error&) {
+                slogger.warn("{}[{}]: Node {} is down for {} verb", desc, uuid(), op_desc, node);
+                nodes_down.emplace(node);
+            } catch (...) {
+                slogger.warn("{}[{}]: Node {} failed {} verb: {}", desc, uuid(), node, op_desc, std::current_exception());
+                nodes_failed.emplace(node);
+            }
+        });
+        std::vector<sstring> errors;
+        if (!nodes_failed.empty()) {
+            errors.emplace_back(format("The {} command failed on nodes={}", op_desc, nodes_failed));
+        }
+        if (!nodes_unknown_verb.empty()) {
+            if (is_prepare) {
+                errors.emplace_back(format("The {} command is unsupported on nodes={}. Please upgrade your cluster and run operation again", op_desc, nodes_unknown_verb));
+            } else {
+                errors.emplace_back(format("The ops_uuid={} was not found or the {} command is unsupported on nodes={}", uuid(), op_desc, nodes_unknown_verb));
+            }
+        }
+        if (!nodes_down.empty()) {
+            errors.emplace_back(format("The {} command failed on nodes={}: the needed nodes are down. It is highly recommended to fix the down nodes and try again", op_desc, nodes_failed));
         }
+        if (!errors.empty()) {
+            co_await coroutine::return_exception(std::runtime_error(boost::algorithm::join(errors, "; ")));
```

Comment:
nit: can we use to_string.hh::join to avoid boost here?

Reply:
Done in v4